persistence/internal/providertest/offsetstore/transaction.go

```diff
@@ -0,0 +1,198 @@
+func DeclareTransactionTests(tc *common.TestContext) {
+	ginkgo.Describe("type offsetstore.Transaction", func() {
+		var (
+			dataStore  persistence.DataStore
+			repository offsetstore.Repository
+			tearDown   func()
+		)
+
+		ginkgo.BeforeEach(func() {
+			dataStore, tearDown = tc.SetupDataStore()
+			repository = dataStore.OffsetStoreRepository()
+		})
+
+		ginkgo.AfterEach(func() {
+			tearDown()
+		})
+
+		ginkgo.Describe("func SaveOffset()", func() {
+			ginkgo.When("application has no previous offsets associated", func() {
+				ginkgo.It("saves an offset with the current offset equal to zero", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+			})
+
+			ginkgo.When("application has previous offsets associated", func() {
+				ginkgo.It("saves an offset with the current offset passed", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					c = n
+					n += n
+
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+
+				ginkgo.It("does not update the current offset with the equal next offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					c = n
+
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+			})
+
+			ginkgo.When("called multiple times within the same transaction", func() {
+				ginkgo.It("saves the last offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+
+					err := persistence.WithTransaction(
+						tc.Context,
+						dataStore,
+						func(tx persistence.ManagedTransaction) error {
+							err := tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
+							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+							c = n
+							n += n
+
+							err = tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
+							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+							return nil
+						},
+					)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+			})
+
+			ginkgo.When("an OCC conflict occurs", func() {
```

Comment:
Some kind of OCC testing also needs to be performed in the `when application has no previous offsets associated` block. I'd retain this table test but move it to the`when application has previous offsets associated` section.